services/api/lib/thumbnails.js

```diff
@@ -72,45 +83,78 @@ exports.register = function(server, options, done) {
   // Check if the given page has the lowest id in its parent project
   // If it is, then request a new thumbnail
   function checkPageId(page, tail) {
-    if ( !process.env.THUMBNAIL_SERVICE_URL ) {
-      return tail();
-    }
-
-    server.methods.pages.min([
-      page.project_id
-    ], function(err, result) {
-      if ( err ) {
-        server.log('error', {
-          details: 'Error querying DB for lowest page ID in project: ' + page.project_id,
-          error: err
-        });
-        return tail(err);
+    process.nextTick(function() {
+      if ( !process.env.THUMBNAIL_SERVICE_URL ) {
+        return tail();
       }
-      var row = result.rows[0];
+      server.methods.pages.min([
+        page.project_id
+      ], function(err, result) {
+        if ( err ) {
+          server.log('error', {
+            details: 'Error querying DB for lowest page ID in project: ' + page.project_id,
+            error: err
+          });
+          return tail(err);
+        }
 
-      if ( row.page_id !== page.id ) {
-        server.debug('Thumbnail update not required');
-        return tail();
+        var row = result.rows[0];
+
+        if ( !row || row.page_id !== page.id ) {
+          server.debug('Thumbnail update not required');
+          return tail();
+        }
+
+        server.debug('Updating thumbnail for project');
+        generateThumbnail(row, tail);
+      });
+    });
+  }
+
+  // We need only check if a thumbnail update is necessary for the lowest pageId in a given set of actions
+  function processBulkThumbnailRequests(request, rawResults) {
+    // reduce this array into a list of thumbnail checks to do
+    rawResults.filter(function(result) {
+      // project updates, and any kind of delete don't trigger thumbnail changes.
+      return result.type !== 'projects' && result.method !== 'remove';
+    }).map(function(result) {
+      var rowData = result.rows[0];
+      var projectId = rowData.project_id;
+      var pageId = rowData.page_id || rowData.id;
+
+      return {
+        project_id: +projectId,
+        page_id: + pageId
+      };
+    }).reduce(function(pendingUpdates, result) {
+      var resultData;
+
+      // iterate over existing thumbnail updates to enable some comparisons
+      for (var i = 0; i < pendingUpdates.length; i++) {
+        resultData = pendingUpdates[i];
+        if (resultData.project_id !== result.project_id) {
+          continue;
+        }
+        if (result.page_id < resultData.page_id) {
+          break;
+        }
+        return pendingUpdates;
       }
 
-      server.debug('Updating thumbnail for project');
-      generateThumbnail(row, tail);
+      pendingUpdates.push(result);
+      return pendingUpdates;
+    }, [])
+    .map(function(result) {
```

Comment:
I think this should be a `forEach` since you're not returning the value here, right?